pex/pex_bootstrapper.py

```diff
@@ -72,33 +72,40 @@ def _iter_interpreters():
         # type: () -> Iterator[InterpreterOrError]
         seen = set()
 
-        paths = None  # type: Optional[MutableSet[str]]
-        if path:
-            paths = OrderedSet(os.path.realpath(p) for p in path.split(os.pathsep))
+        normalized_paths = (
+            OrderedSet(os.path.realpath(p) for p in path.split(os.pathsep)) if path else None
+        )
 
+        # Check if the current interpreter is valid. If `path` was specified, the current
+        # interpreter must be included in the value; otherwise, we always yield the interpreter.
         current_interpreter = PythonInterpreter.get()
         if not _valid_path or _valid_path(current_interpreter.binary):
-            if paths:
-                # Prefer the current interpreter if present on the `path`.
+            if normalized_paths:
                 candidate_paths = frozenset(
                     (current_interpreter.binary, os.path.dirname(current_interpreter.binary))
                 )
-                candidate_paths_in_path = candidate_paths.intersection(paths)
+                candidate_paths_in_path = candidate_paths.intersection(normalized_paths)
                 if candidate_paths_in_path:
                     for p in candidate_paths_in_path:
-                        paths.remove(p)
+                        normalized_paths.remove(p)
```

Comment:
I know this was the case before as well, but I think I see a problem here: we remove not just the full path to the interpreter, but also the directory containing it. What if there are other interpreters to consider in that directory?

Reply:
Also, since we do the `seen` check above, why do we even bother removing anything from `normalized_paths`?

Reply:
> why do we even bother removing anything from normalized_paths?

Good question. I suspect performance? @jsirois do you remember the motivation in https://github.com/pantsbuild/pex/pull/783?

Reply:
Not doing that would remove this weird case, no?

Reply:
Indeed. If we don't remove, we avoid this edge case without needing any special `if` statement and confusing comment.

Reply:
I fixed this another way, but we may still want to remove this code. 

I think I'm not comfortable removing it until we hear back from John, though, so let's save for a followup? Iirc, he's said in the past that interpreter discovery is expensive, hence optimizations like preferring the current interpreter.

Reply:
I still think this is actually wrong though. We remove the directory, when there may be other interpreters in it.

Reply:
Still, that is the current behavior, so we can fix in a followup.

Reply:
This is definitely a bug: #1109 

That said, the if statement is still needed. If the function is called with a non `None` `path` then we must respect that and only include the current interpreter if it is a member of that `path`.